diff --git a/target/microblaze/translate.c b/target/microblaze/translate.c
index 9325661785..13fdd40284 100644
--- a/target/microblaze/translate.c
+++ b/target/microblaze/translate.c
@@ -1610,6 +1610,32 @@ static inline void decode(DisasContext *dc, uint32_t ir)
             cpu_abort(CPU(dc->cpu), "fetching nop sequence\n");
         }
     }
+
+   if(dc->cpu->env.sregs[SR_PC] == 0x08) {
+	   cpu_abort(CPU(dc->cpu), "Abort, SW Exception\n");
+   }
+   else if(dc->cpu->env.sregs[SR_PC] == 0x10) {
+	   cpu_abort(CPU(dc->cpu), "Abort, Interrupt\n");
+   }
+   else if(dc->cpu->env.sregs[SR_PC] == 0x20) {
+   	   cpu_abort(CPU(dc->cpu), "Abort, HW Exception\n");
+   }
+
+//   fprintf(stdout,"PC = %X\n", dc->cpu->env.regs[SR_PC]);
+
+    if(ir == 0xB8000000){ // bri 0
+    	MicroBlazeCPU *cpu = dc->cpu;
+    	CPUMBState *env = &cpu->env;
+    	unsigned int ret_val = env->regs[5];
+    	fprintf(stdout,"Exit code : %d\n",ret_val);
+    	if(ret_val == 0){
+    		exit(0);
+    	}
+    	else {
+    		cpu_abort(CPU(cpu), "Abort, exit code = %d\n", ret_val);
+    	}
+    }
+
     /* bit 2 seems to indicate insn type.  */
     dc->type_b = ir & (1 << 29);
 
